### DISCUSSION TITLE
Basic patterns

### DISCUSSION THREAD

**Discussion by ewal** [excaliburjs/Excalibur#2362] [Q&A] [answered]
Hello there,

Been working as a developer for many years but have never built a game so I thought it was about time to give it a try. I fell for the Excalibur platform since it's typed plus I'm familiar with the stack in general. There are however many concepts to familiarize with that are completely new to me that I very much do enjoy reading and learning about, but, progress is quite slow for different reasons.

What I'm basically curious about right now is what patterns you work with to build games that can scale or if there are general patterns to apply when building games using the Excalibur platform? Still a super n00b so I'm grateful and appreciate all help and pointers I can get. (Not yet quite sure what questions to ask either :))

**Comment by eonarheim** [comment, marked as answer]
Hi @ewal, thanks for the kind words!

As far as patterns go, we've been pretty pleased with how the structure of our latest jam game turned out. 

https://github.com/excaliburjs/ludum-50

What we've found that works well generally is the following:
* Keep the main.ts entrypoint as small as possible (this is where we create a `new ex.Engine()` and configure anything else at the engine level)
* Keep all of our assets/resources in one file `resources.ts` and load them all at once in the main
* Keep a `config.ts` for tweaking global constant values easily, we've wired up things like `dat.gui` or `tweakpane` that make changing config during development easy
* Where possible we use `onInitialize()` for initialization logic over the constructor, this saves CPU cycles because it's called before the first update an entity is needed  and makes it easy to restart a game or reusing something by manually calling `onInitialize()`.
* Use `ex.Scene` (like in `level.ts`) `onInitialize()` as a composition root for your game to assemble all the different bits
* We've found extending `ex.Actor` is a useful way to create interactable game objects, a descriptive file name per thing makes it easy to find.

Hope this helps! I'll probably spend some more time on these and put them into our docs as well 👍

**Reply by eonarheim** [reply]
Let me know if there are some specifics or different areas you'd like to dig into as well!

**Comment by ewal** [comment]
Thank you for your reply! I appreciate it very much.
I'll study the code in the ludum-50 repo, thanks.

I will return soon with some more specific questions and a repo with some concrete examples :)
